### PR TITLE
Fix "use after closing" error when calling pool.getResource()

### PR DESCRIPTION
**Pull request by william-cheung**
There exists race condition between pool.close() and pool.getResource(), which may cause the following exception：
   redis.clients.jedis.exceptions.JedisConnectionException: Could not get a resource from the pool
	at redis.clients.util.Pool.getResource(Pool.java:53)
	at redis.clients.jedis.JedisPool.getResource(JedisPool.java:226)
	at io.codis.jodis.RoundRobinJedisPool.getResource(RoundRobinJedisPool.java:220)
	...

I proposed two solutions to this problem:
   1. defer "close" of a pool; 2. close the pool when PooledObject.finalize() is called
   
The first solution may cause errors under conditions that rarely happen, but it is more deterministic than the second on resource collection. I prefer the first one.

### CODE REVIEW

**Comment by william-cheung** [comment]
This fix will enable rolling-restart of codis proxies without affecting codis clients. The rolling-restart process should be like:

```python
for proxy in proxies:
      mark_offline(proxy)        # set the status of 'proxy' to 'offline'
      sleep(10) .                       # wait 10 seconds for all ongoing requests to 'proxy' completed
      restart(proxy)                 # kill 'proxy' and restart it
```

File: src/main/java/io/codis/jodis/RoundRobinJedisPool.java [hunk 1]
```diff
@@ -96,6 +96,21 @@ public PooledObject(String addr, JedisPool pool) {
...
+            return pool.getResource();
+        }
+
+        public void close() {
+            pool.close();
+            LOG.info(String.format("Connection pool to %s closed", addr));
+        }
+
+//        @Override
+//        protected void finalize() throws Throwable {
```
Comment (Apache9) [inline src/main/java/io/codis/jodis/RoundRobinJedisPool.java]: Just remove this method.

File: src/main/java/io/codis/jodis/RoundRobinJedisPool.java [hunk 2]
```diff
@@ -199,9 +214,20 @@ private void resetPools() {
             }
         }
         this.pools = builder.build();
-        for (PooledObject pool: addr2Pool.values()) {
+        for (final PooledObject pool: addr2Pool.values()) {
             LOG.info("Remove proxy: " + pool.addr);
-            pool.pool.close();
+            new Thread(new Runnable() {
```
Comment (Apache9) [inline src/main/java/io/codis/jodis/RoundRobinJedisPool.java]: Better using ScheduledThreadPool?

Comment (william-cheung) [reply src/main/java/io/codis/jodis/RoundRobinJedisPool.java]: Yes, ScheduledThreadPoolExecutor is better. ScheduledThreadPoolExecutor uses a priority queue to maintain tasks. The algorithm is not optimal, but it's ok for our case.

**Pull request closed without merging**